### DISCUSSION TITLE
UserControl and TemplateControl

### DISCUSSION THREAD

**Discussion by GoldenGames18** [AvaloniaUI/Avalonia#17068] [Q&A]
Hello, I have been using Avalonia for several days now to work on a personal project, and I have encountered an issue understanding UserControl and TemplateControl. After searching through various tutorials, I would like to know if it is possible to create reusable components that could modify data in my ViewModel. If so, could you show me a concrete example? In my case, I need to create an input to encode an IP address using four TextBoxes.

**Comment by thevortexcloud** [comment]
>  I would like to know if it is possible to create reusable components that could modify data in my ViewModel. 

It is. Some of Avalonia's docs are somewhat sparse. Luckily a lot of Avalonia is very similar to WPF, so the WPF examples generally "just work" (albeit with some minor tweaks, [such as changing `DepedendencyProperty` to `StyledProperty`](https://docs.avaloniaui.net/docs/get-started/wpf/dependencyproperty)).

https://learn.microsoft.com/en-us/dotnet/api/system.windows.controls.usercontrol?view=windowsdesktop-8.0

Whether or not you should use a user control or a template control really depends on where and how you want to use it. If it's just for your app only, you can probably get away with a user control.

That said controls generally don't care about view models at all. All they really need to know about are bindings to view models.
